File: Week_06/1143/longest_common_subsequence3.go
package main

func longestCommonSubsequence3(text1 string, text2 string) int {
	m, n := len(text1), len(text2)
	dp := make([]int, n+1)
	for i := 1; i <= m; i++ {
		pre := 0
		for j := 1; j <= n; j++ {
			tmp := dp[j]
			if text1[i-1] == text2[j-1] {
				dp[j] = pre + 1
			} else {
				dp[j] = max(dp[j-1], dp[j])
			}
			pre = tmp
		}
	}
	return dp[n]
}

File: Week_06/1143/longest_common_subsequence3.py
class Solution:
    def longestCommonSubsequence(self, text1: str, text2: str) -> int:
        m, n = len(text1), len(text2)

        dp = [0] * (n + 1)

        for i in range(1, m + 1):
            pre = 0
            for j in range(1, n + 1):
                tmp = dp[j]
                if text1[i - 1] == text2[j - 1]:
                    dp[j] = pre + 1
                else:
                    dp[j] = max(dp[j], dp[j - 1])
                pre = tmp
        return dp[n]
